src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java

```diff
@@ -0,0 +1,881 @@
+        }
+
+        /**
+         * Checks whether the current tool state is according to this count-column manager
+         * instance requirements.
+         * <p>
+         *     Typically this means that components in the tool that contain relevant information
+         *     for this count-column manager, have been initialized (e.g. the {@link #sampleDB}).
+         * </p>
+         *
+         * @param tool the tool to test.
+         * @throws IllegalArgumentException if {@code tool} is {@code null}.
+         * @throws IllegalStateException if the current state of {@code tool} is not compatible with
+         *    this count-column manager type.
+         */
+        protected void checkToolState(final ExomeReadCounts tool) {
+            if (tool == null) {
+                throw new IllegalArgumentException("the tool cannot be null");
+            }
+        }
+
+        /**
+         * Returns the name for a column given its index.
+         *
+         * <p>
+         *     Indexes are 0-based.
+         * </p>
+         * @param index the column index.
+         * @return never {@code null}.
+         */
+        protected abstract String columnName(final int index);
+
+        /**
+         * Composes the list of count column names.
+         *
+         * <p>
+         *     The result list would have exactly <code>{@link #columnCount()}</code> elements.
+         * </p>
+         *
+         * @return never {@code null}.
+         */
+        private List<String> composeCountColumns() {
+            final int columnCount = columnCount();
+            final Set<String> resultSet = new LinkedHashSet<>(columnCount);
+            final String[] result = new String[columnCount];
+            for (int i = 0; i < columnCount; i++) {
+                resultSet.add(result[i] = columnName(i));
+            }
+            if (resultSet.size() != columnCount) {
+                throw new UserException.BadArgumentValue(COUNT_COLUMN_FULL_NAME,
+                        "count column name pattern results in more than one column with the same name: " +
+                                String.join(", ", result));
+            }
+            return Collections.unmodifiableList(Arrays.asList(result));
+        }
+
+        /**
+         * Returns the count column index corresponding to a read.
+         * @param read the query read.
+         * @return a value from 0 to <code>{@link #columnCount()} - 1</code>, or -1 if the value is to be discarded.
+         */
+        protected abstract int columnIndex(SAMRecord read);
+
+        /**
+         * Returns the count column headers.
+         * @return never {@code null}, an immutable list of header where
+         */
+        private List<String> columnNames() {
+            return columnNames;
+        }
+
+        /**
+         * Returns the number of count columns headers.
+         * @return never {@code null}.
+         */
+        protected abstract int columnCount();
```

Comment:
This method is tricky and may lead to problems. The only time it's called is from inside of `composeCountColumns` and so the natural implementation of `columnCount` will lead to a NPE (the natural implementation being `columnNames.size()`. How about making a field for it and store in in the constructor (you know the size at that time).